django/contrib/admin/options.py

```diff
@@ -306,7 +306,16 @@ def get_readonly_fields(self, request, obj=None):
         """
         Hook for specifying custom readonly fields.
         """
-        return self.readonly_fields
+        if hasattr(request, 'user') and (self.has_change_permission(request, obj) or self.has_add_permission(request)):
```

Comment:
Putting this logic in the default implementation of `get_readonly_fields()` could be problematic. Anyone overriding the default implementation will no longer be able to use the read-only admin, correct?